bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java

```diff
@@ -172,9 +173,33 @@ public boolean tableExists(String tableName) throws IOException {
     return response;
   }
 
+  // Workaround for #2016, BigtableTableAdminClient#getTable throws TableNotFoundException(due to
+  // two step details fetch) and an intermediate exception(FAILED_PRECONDITION).
+  //
+  // This method is only being called when user sends `null` or `""` string to listTable(Pattern)
+  // which in being used by deleteTables(), enableTables(), disableTables(),
+  // getTableDescriptorByTableNames().
+  private HTableDescriptor[] getTableDescriptorsIgnoreFailure(TableName[] tableNames) {
+    List<HTableDescriptor> descriptors = new ArrayList<>();
+    for (TableName tableName : tableNames) {
+      try {
+        descriptors.add(getTableDescriptor(tableName));
+      } catch (IOException ex) {
+        // Do we need better error message around it.
+        LOG.warn(
+            "Table name %s failed with %s with error message as %s",
+            tableName.getNameAsString(), ex.getClass().getSimpleName(), ex.getMessage());
+      }
+    }
+    return descriptors.toArray(new HTableDescriptor[0]);
+  }
+
   /** {@inheritDoc} */
   @Override
   public HTableDescriptor[] listTables(Pattern pattern) throws IOException {
+    if (pattern == null) {
+      return getTableDescriptorsIgnoreFailure(listTableNames());
+    }
     // NOTE: We don't have systables.
     return getTableDescriptors(listTableNames(pattern));
```

Comment:
Why shouldn't this suppress the errors from the race condition?

Reply:
I thought to keep `getTableDescriptors()` as it is. So far no user has logged any issue for `listTables(Pattern)`. Another thought behind this was If it is valid pattern beside `".*"` then it would certainly fetch small of a number of tables.

If you think we should guard getTableDescriptors() for race condition then I can raise separate PR considering that is a Public method.

Reply:
It still vulnerable to the same race conditions. Please fix this.

getTableDescriptors(tableNames) is different in that the caller is asking for specific tables, so it should throw an error.

Reply:
I believe by race condition you mean when listTables() is fetching a large number of TableDescriptor using getTable(). 

Should this be an acceptable(ignoring failure even if the user has provided prefix/regex):
```java
  @Override
  public HTableDescriptor[] listTables() throws IOException {
    // NOTE: We don't have systables.
    return getTableDescriptorsIgnoreFailure(listTableNames());
  }
```